### ISSUE TITLE
Some `orchest-cli` commands should not require an active cluster

### ISSUE THREAD

**Issue by astrojuanlu** [orchest/orchest#1053]
**Describe the problem this improvement solves**
I tried calling `orchest version --latest` in development, but got

```
$ orchest version --latest
Traceback (most recent call last):
  File "/home/juanlu/Projects/Orchest/orchest/orchest-cli/orchestcli/cmds.py", line 30, in <module>
    config.load_kube_config()
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/kube_config.py", line 808, in load_kube_config
    loader = _get_kube_config_loader(
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/kube_config.py", line 770, in _get_kube_config_loader
    return KubeConfigLoader(
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/kube_config.py", line 206, in __init__
    self.set_active_context(active_context)
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/kube_config.py", line 259, in set_active_context
    self._current_context = self._config['contexts'].get_with_name(
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/kube_config.py", line 657, in get_with_name
    raise ConfigException(
kubernetes.config.config_exception.ConfigException: Invalid kube-config file. Expected object with name  in /home/juanlu/.kube/config/contexts list

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/juanlu/Projects/Orchest/orchest/.venv/bin/orchest", line 33, in <module>
    sys.exit(load_entry_point('orchest-cli', 'console_scripts', 'orchest')())
  File "/home/juanlu/Projects/Orchest/orchest/.venv/bin/orchest", line 25, in importlib_load_entry_point
    return next(matches).load()
  File "/usr/lib/python3.9/importlib/metadata.py", line 86, in load
    module = import_module(match.group('module'))
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1030, in _gcd_import
  File "<frozen importlib._bootstrap>", line 1007, in _find_and_load
  File "<frozen importlib._bootstrap>", line 986, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 680, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 850, in exec_module
  File "<frozen importlib._bootstrap>", line 228, in _call_with_frames_removed
  File "/home/juanlu/Projects/Orchest/orchest/orchest-cli/orchestcli/cli.py", line 57, in <module>
    from orchestcli import cmds
  File "/home/juanlu/Projects/Orchest/orchest/orchest-cli/orchestcli/cmds.py", line 32, in <module>
    config.load_incluster_config()
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/incluster_config.py", line 118, in load_incluster_config
    InClusterConfigLoader(
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/incluster_config.py", line 54, in load_and_set
    self._load_config()
  File "/home/juanlu/Projects/Orchest/orchest/.venv/lib/python3.9/site-packages/kubernetes/config/incluster_config.py", line 62, in _load_config
    raise ConfigException("Service host/port is not set.")
kubernetes.config.config_exception.ConfigException: Service host/port is not set.
```

However, after doing `minikube start` (even with Orchest stopped), it works:

```
$ orchest version --latest
v2022.06.5
```

**Describe the solution you'd like**
Only the `orchest-cli` commands that need communication with the cluster require an active cluster.

In addition, we could catch this error and provide a nicer looking one?

**Comment by yannickperrenet** [comment]
Good thing that you caught this error. It almost makes me want to remove the `orchest version --latest` command, because it is the only command/flag that does not require a cluster. The purpose of the `orchest-cli` is to manage Orchest on a running Kubernetes cluster and `orchest version --latest` was just added a convenience method. 

> In addition, we could catch this error and provide a nicer looking one?

This we can definitely do! 

In summary, not sure what to do with `orchest version --latest` :thinking:

**Comment by yannickperrenet** [comment]
In short, what we could do is loading the kube config (`config.load_kube_config()`) on initialization of `kubernetes.client` classes (e.g. `ApiClient()`) and initialize a new client class **every time** we want to make a k8s API call. 

Longer version. On initialization of `kubernetes.client` classes the `api_client.ApiClient()` is initialized which in turn will try to load the kube config from the `Configuration` singleton (and so `config.load_kube_config()` needs to be called before initialization of these client classes). To achieve this we can decorate all the client classes we use to call `config.load_kube_config()` (or do some smart caching using the `Configuration` singleton) before `__init__(...)` is invoked and invoke `__init__(...)` as late as possible. Example:
```python
def run_first_lazily(func):
    def inner(*args, **kwargs):
        print("Loading kube config")
        return func(*args, **kwargs)

    return inner

def class_run_first_lazily(cls):
    def inner(*args, **kwargs):
        for attr_name, attr in vars(cls).items():
            if callable(attr) and attr_name == "__init__":
                setattr(cls, attr_name, run_first_lazily(attr))
        return cls(*args, **kwargs)  # passed to cls.__init__

    return inner

# Usage in `orchest-cli/orchestcli/cmds.py` would become
CORE_API = class_run_first_lazily(client.CoreV1Api)  # instead of `CORE_API = client.CoreV1Api()`
...
CORE_API().delete_namespace(ns)  # instead of `CORE_API.delete_namespace(ns)`
```
In the example you can see that we initialize the class only the moment we actually make an API call.

**Pros**
* We can now have CLI commands that do not require a valid kube config, e.g. `orchest version --latest`

**Cons**
* Every future implementer of functionality in the CLI needs to make sure that these client classes are initialized at the last possible moment as to adhere to this new pattern (we can of course make a note of this in the module docstring).
* On initialization of the client classes a deepcopy of the configuration is made and thus by initialization for every API call instead of just once there will be a performance hit (although rather minimal I assume but that depends on the size of the kube config). In addition, although highly unlikely, the kube config could have changed between API calls and thus by loading the kube config again the subsequent API calls could be made to other clusters.

---
In conclusion, I am not sure going this route is worth it just for the `orchest version --latest` command to not require a cluster. As for the `kubectl` installation docs, we could just leave the `curl` command (as per https://github.com/orchest/orchest/commit/5e9e62452b6a66599a8c95af7223d8f3c48e1873#commitcomment-76699279).

**Comment by yannickperrenet** [comment]
> In addition, we could catch this error and provide a nicer looking one?

This is now solved. 

Feel free to re-open is you disagree with the consensus in https://github.com/orchest/orchest/issues/1053#issuecomment-1164085100 :wink:

**Comment by ricklamers** [comment]
To support https://github.com/orchest/orchest/issues/1053#issuecomment-1164085100, you can just run `pip show orchest-cli` for the CLI version.

**Comment by yannickperrenet** [comment]
> you can just run `pip show orchest-cli` for the CLI version.

Indeed if you don't have a cluster running then you can fall back to `pip show orchest-cli` instead of `orchest --version`, which is different from `orchest version` (get's the current running Orchest version on the cluster) and `orchest version --latest` (get's the latest Orchest version that is released) :+1:
